app/Services/Doctor/DoctorVaccinationService.cs

```diff
@@ -0,0 +1,90 @@
+﻿using backend.Database;
+using backend.Dto.Responses;
+using backend.Exceptions;
+using backend.Helpers;
+using backend.Models.Accounts;
+using backend.Models.Visits;
+
+namespace backend.Services.Doctor
+{
+    public class DoctorVaccinationService
+    {
+        private readonly DataContext dataContext;
+        private readonly Mailer mailer;
+
+        public DoctorVaccinationService(DataContext dataContext, Mailer mailer)
+        {
+            this.dataContext = dataContext;
+            this.mailer = mailer;
+        }
+
+        public async Task<SuccessResponse> ChangeVaccinationSlot(int vaccinationId, int newVaccinationSlotId, DoctorModel doctor)
+        {
+            // Find vaccination visit with matching id
+            VaccinationModel vaccination = this.dataContext
+                .Vaccinations
+                .FirstOrThrow(vaccination => vaccination.Id == vaccinationId, new NotFoundException("Vaccination visit not found"));
+
+            // Check if vaccination visit belongs to editing doctor
+            if(vaccination.Doctor != doctor)
+            {
+                throw new ConflictException("Vaccination visit is assigned to another doctor.");
+            }
+
+            // Get slot connected to found vaccination visit
+            VaccinationSlotModel currentSlot = this.dataContext
+                .VaccinationSlots
+                .FirstOrThrow(slot => slot.Id == vaccination.VaccinationSlotId, new NotFoundException("Current vaccination slot not found"));
+
+            // Get new slot
+            VaccinationSlotModel newSlot = this.dataContext
+                .VaccinationSlots
+                .FirstOrThrow(slot => slot.Id == newVaccinationSlotId, new NotFoundException("New vaccination slot not found"));
+
+            // Block concurrent access
+            Semaphores.slotSemaphore.WaitOne();
```

Comment:
same here, maybe move it up before searching for the new slot

Reply:
Fixed 👍